cep-recipe-jinja.md

```diff
@@ -0,0 +1,511 @@
+take arguments, such as `... | replace('foo', 'bar')`.
+
+The following Jinja filters are available, taken from the upstream `minijinja`
+library:
+
+- `replace`: replace a string with another string (e.g. `"{{ 'foo' | replace('oo', 'aa') }}"` will return `"faa"`)
+- `lower`: convert a string to lowercase (e.g. `"{{ 'FOO' | lower }}"` will return `"foo"`)
+- `upper`: convert a string to uppercase (e.g. `"{{ 'foo' | upper }}"` will
+return `"FOO"`) - `int`: convert a string to an integer (e.g. `"{{ '42' | int }}"` will return `42`)
+- `abs`: return the absolute value of a number (e.g. `"{{ -42 | abs }}"` will return `42`)
+- `bool`: convert a value to a boolean (e.g. `"{{ 'foo' | bool }}"` will return `true`)
+- `default`: return a default value if the value is falsy (e.g. `"{{ '' | default('foo') }}"` will return `"foo"`)
+- `first`: return the first element of a list (e.g. `"{{ [1, 2, 3] | first }}"`
+will return `1`) - `last`: return the last element of a list (e.g. `"{{ [1, 2, 3] | last }}"` will return `3`)
+- `length`: return the length of a list (e.g. `"{{ [1, 2, 3] | length }}"` will return `3`)
+- `list`: convert a string to a list (e.g. `"{{ 'foo' | list }}"` will return `['f', 'o', 'o']`)
+- `join`: join a list with a separator (e.g. `"{{ [1, 2, 3] | join('.') }}"` will return `"1.2.3"`)
+- `min`: return the minimum value of a list (e.g. `"{{ [1, 2, 3] | min }}"` will return `1`)
+- `max`: return the maximum value of a list (e.g. `"{{ [1, 2, 3] | max }}"` will return `3`)
+- `reverse`: reverse a list (e.g. `"{{ [1, 2, 3] | reverse }}"` will return `[3, 2, 1]`)
+- `slice`: slice a list (e.g. `"{{ [1, 2, 3] | slice(1, 2) }}"` will return `[2]`)
+- `batch`: This filter works pretty much like `slice` just the other way round.
+  It returns a list of lists with the given number of items. If you provide a
+  second parameter this is used to fill up missing items.
+- `sort`: sort a list (e.g. `"{{ [3, 1, 2] | sort }}"` will return `[1, 2, 3]`)
+- `trim`: remove leading and trailing whitespace from a string (e.g. `"{{ ' foo ' | trim }}"` will return `"foo"`)
+- `unique`: remove duplicates from a list (e.g. `"{{ [1, 2, 1, 3] | unique }}"` will return `[1, 2, 3]`)
+- `split`: split a string into a list (e.g. `"{{ '1.2.3' | split('.') }}"` will return `['1', '2', '3']`). By default, splits on whitespace.
+
+<details>
+<summary>Removed filters</summary>
+The following filters are removed from the builtins:
+
+- `attr`
+- `indent` (indent with spaces, could be useful?)
+- `select`
+- `selectattr`
+- `dictsort`
+- `reject`
+- `rejectattr`
+- `round`
+- `map`
+- `title`
+- `capitalize`
+- `urlencode`
+- `escape`
+- `pprint`
+- `safe`
+- `items`
+- `float`
+- `tojson`
+</details>
+
+
+### Extra filters for recipes
+
+#### The `version_to_buildstring` filter
+
+- `${{ python | version_to_buildstring }}` converts a version from the variant
+  to a build string (it removes the `.` character and takes only the first two
+  elements of the version).
```

Comment:
I'm not so sure about the utility of this one.
Yes, this how we currently compose parts of build strings.
But it's also a hacky thing to do, really.
This might just support dragging on such build string massaging for longer than we should.

(That is not to say that I'm completely opposed to such functionality; I just want to caution against adding potential future burden from the get go.)

Reply:
Yes, it'd be a `python.split(".")[:2] | join("")` -- so, not too complicated, but also error-prone.
My remark was more about why we do these stuff in general, though.
I think turning a `11.2.0` into `112` is just not wise but just a historic habit that we might want to weed out...

Reply:
Actually, this syntax doesn't work in the new recipes (at least not in the way implemneted in `rattler-build` since we're using `minijinja` that doesn't have all the Python builtins implemented).

Reply:
using `''` as the join seperator is almost certainly a bad idea even though we do it all the time.  `_` is a FAR better solution since it avoids the `1.12.0` == `11.2.0` problem that this has